simulator/infra/cache/cache_tag_array.cpp

```diff
@@ -120,7 +123,60 @@ uint32 CacheTagArray::write( Addr addr)
 
     // get cache coordinates
     const uint32 num_set = set( addr);
-    const auto way = narrow_cast<uint32>( replacement_module.update( num_set));
+    const auto way = narrow_cast<uint32>( replacement_module->update( num_set));
+
+    // get an old tag
+    auto& entry = tags[ num_set][ way];
+    const Addr old_tag = entry.tag;
+
+    // Remove old tag from lookup helper and add a new tag
+    if ( entry.is_valid)
+        lookup_helper[ num_set].erase( old_tag);
+    lookup_helper[ num_set].emplace( new_tag, way);
+
+    // Update tag array
+    entry.tag = new_tag;
+    entry.is_valid = true;
+
+    return way;
+}
+
+AlwaysHitCacheTagArray::AlwaysHitCacheTagArray(
+    uint32 size_in_bytes,
+    uint32 ways,
+    uint32 line_size,
+    uint32 addr_size_in_bits)
+    : CacheTagArray( size_in_bytes, ways, line_size, addr_size_in_bits)
```

Comment:
Please see my previous comment about hierarchy.
Inheritance creates a "is-a" relationship between classes. For instance, if class Triangle is inherited from class Shape, that means "Triangle is a Shape".
Here you inherit InfiniteCacheTagArray from FiniteCacheTagArray implementation, which makes non-sense. That architectural mistake leads to questionable implementation things, I highlighted them below.